In [31]:
import findspark
findspark.init()

In [32]:
import pyspark # only run after findspark.init()
import pandas as pd
from pyspark.sql import SparkSession
import csv

In [33]:
sc = SparkSession.builder.master("local").getOrCreate()
# sc = SparkSession.builder.master("spark://cheyenne:30915").getOrCreate()

Read the data into a pyspark dataframe.

In [34]:
greenhouseGasInitialData = sc.read.csv("owid-co2-data.csv")

Rename and drop unneeded columns per our feature selection decisions.

In [35]:
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c0", "iso_code")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c1", "country")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c2", "year")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c3", "co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c4", "co2_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c5", "trade_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c6", "cement_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c7", "cement_co2_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c8", "coal_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c9", "coal_co2_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c10", "flaring_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c11", "flaring_co2_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c12", "gas_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c13", "gas_co2_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c14", "oil_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c15", "oil_co2_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c16", "other_industry_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c17", "other_co2_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c18", "co2_growth_prct")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c19", "co2_growth_abs")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c20", "co2_per_gdp")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c21", "co2_per_unit_energy")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c22", "consumption_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c23", "consumption_co2_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c24", "consumption_co2_per_gdp")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c25", "cumulative_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c26", "cumulative_cement_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c27", "cumulative_coal_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c28", "cumulative_flaring_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c29", "cumulative_gas_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c30", "cumulative_oil_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c31", "cumulative_other_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c32", "trade_co2_share")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c33", "share_global_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c34", "share_global_cement_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c35", "share_global_coal_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c36", "share_global_flaring_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c37", "share_global_gas_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c38", "share_global_oil_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c39", "share_global_other_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c40", "share_global_cumulative_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c41", "share_global_cumulative_cement_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c42", "share_global_cumulative_coal_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c43", "share_global_cumulative_flaring_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c44", "share_global_cumulative_gas_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c45", "share_global_cumulative_oil_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c46", "share_global_cumulative_other_co2")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c47", "total_ghg")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c48", "ghg_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c49", "total_ghg_excluding_lucf")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c50", "ghg_excluding_lucf_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c51", "methane")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c52", "methane_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c53", "nitrous_oxide")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c54", "nitrous_oxide_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c55", "population")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c56", "gdp")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c57", "primary_energy_consumption")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c58", "energy_per_capita")
greenhouseGasInitialData = greenhouseGasInitialData.withColumnRenamed("_c59", "energy_per_gdp")

In [36]:
greenhouseGasInitialData = greenhouseGasInitialData.drop("iso_code", "co2_per_capita", "trade_co2", "cement_co2", "cement_co2_per_capita", "coal_co2", 
                    "coal_co2_per_capita", "flaring_co2", "flaring_co2_per_capita", "gas_co2", "gas_co2_per_capita",
                   "oil_co2", "oil_co2_per_capita", "other_industry_co2", "other_co2_per_capita",
                   "co2_growth_prct", "co2_growth_abs", "co2_per_gdp", "co2_per_unit_energy", "consumption_co2",
                   "consumption_co2_per_capita", "consumption_co2_per_gdp", "cumulative_cement_co2",
                   "cumulative_coal_co2", "cumulative_flaring_co2", "cumulative_gas_co2", "cumulative_oil_co2",
                   "cumulative_other_co2", "trade_co2_share", "share_global_co2", "share_global_cement_co2",
                   "share_global_coal_co2", "share_global_flaring_co2", "share_global_gas_co2", "share_global_oil_co2",
                   "share_global_other_co2", "share_global_cumulative_co2", "share_global_cumulative_cement_co2", 
                    "share_global_cumulative_coal_co2",
                   "share_global_cumulative_flaring_co2", "share_global_cumulative_gas_co2", "share_global_cumulative_oil_co2",
                    "share_global_cumulative_other_co2",
                   "ghg_per_capita", "total_ghg_excluding_lucf", "ghg_excluding_lucf_per_capita",
                   "methane_per_capita", "nitrous_oxide_per_capita",
                   "population", "gdp", "energy_per_capita", "energy_per_gdp", "primary_energy_consumption", 
                    "iso_code", "total_ghg")

In [37]:
type(greenhouseGasInitialData)

pyspark.sql.dataframe.DataFrame

In [38]:
greenhouseGasInitialDataFiltered = greenhouseGasInitialData.filter("year > 1998")
greenhouseGasInitialDataFiltered = greenhouseGasInitialDataFiltered.filter("year < 2022")

In [39]:
#get rid of none values
greenhouseGasInitialDataFiltered = greenhouseGasInitialDataFiltered.filter("co2 != 'None'")
greenhouseGasInitialDataFiltered = greenhouseGasInitialDataFiltered.filter("cumulative_co2 != 'None'")
greenhouseGasInitialDataFiltered = greenhouseGasInitialDataFiltered.filter("methane != 'None'")
greenhouseGasInitialDataFiltered = greenhouseGasInitialDataFiltered.filter("nitrous_oxide != 'None'")

In [40]:
greenhouseGasInitialDataFiltered.count()

3897

In [12]:
greenhouseGasInitialDataFilteredCollected = greenhouseGasInitialDataFiltered.collect()

We want to match these greenhouse gases up with the weather data set. The north region Brazilian weather data set starts at 2000-05 (May of 2000). There is no data from 1999 in the weather dataset, but there is data up to and including 2021-04. There is no data for 2022.

There is no data in the GHG dataset for 2021 or 2022.
There is no data for methane or nitrous oxide after 2018, but it does have 2018.
Therefore, we essentially gather data points for 2000-2018 and then extrapolate 2019, 2020, and 2021 to match up with the weather datasets.

In [14]:
greenhouseGasDataFilteredUnEdited = greenhouseGasInitialDataFiltered

In [15]:
averagesInListFormat1999To2018 = []
year = 1999
while year < 2019:

    greenhouseGasInitialDataFilteredCollected = greenhouseGasDataFilteredUnEdited
    greenhouseGasInitialDataFilteredCollected = greenhouseGasInitialDataFilteredCollected.filter(greenhouseGasInitialDataFilteredCollected.year.contains(str(year)))
    averagesInListFormat1999To2018.append(greenhouseGasInitialDataFilteredCollected.agg({'co2': 'avg', 'cumulative_co2': 'avg', 'methane': 'avg', 'nitrous_oxide': 'avg'}).collect()[0])

    year+=1

In [225]:
averagesInListFormat1999To2018 

[Row(avg(cumulative_co2)=11511.07836082474, avg(co2)=265.89206701030935, avg(nitrous_oxide)=27.672010309278345, avg(methane)=74.93850515463916),
 Row(avg(cumulative_co2)=11784.229247422683, avg(co2)=273.1509329896907, avg(nitrous_oxide)=27.528350515463917, avg(methane)=74.6510824742268),
 Row(avg(cumulative_co2)=12059.934247422683, avg(co2)=275.7049793814432, avg(nitrous_oxide)=27.730670103092788, avg(methane)=74.49737113402061),
 Row(avg(cumulative_co2)=12278.275933333334, avg(co2)=280.1875025641026, avg(nitrous_oxide)=28.108769230769237, avg(methane)=75.9751794871795),
 Row(avg(cumulative_co2)=12572.25496923077, avg(co2)=293.97903589743584, avg(nitrous_oxide)=28.420358974358972, avg(methane)=76.2288717948718),
 Row(avg(cumulative_co2)=12878.802599999995, avg(co2)=306.54758974358964, avg(nitrous_oxide)=29.197538461538468, avg(methane)=78.12897435897438),
 Row(avg(cumulative_co2)=13194.991758974358, avg(co2)=316.1891794871795, avg(nitrous_oxide)=29.238307692307693, avg(methane)=77.9642

We write the data to an intermediate csv (which we won't turn in unless requested).
This just makes it easy to bring the data in as tuples.

In [226]:
with open('greenHouseGasAveragesFeatures.csv', 'w', newline='') as file:
    write = csv.writer(file)
    
    for row in averagesInListFormat1999To2018:
            write.writerow([row[0], row[1], row[2], row[3]])

In [17]:
with open('greenHouseGasAveragesFeatures.csv', newline='') as readfile:
    reader = csv.reader(readfile)
    dataFromCSV = tuple(reader)

In [18]:
dataFromCSV = [[float(line[0]), float(line[1]), float(line[2]), float(line[3])] for line in dataFromCSV]

In [24]:
len(dataFromCSV)

20

In the following cells, we take the data for 20 years, from 1999 to 2018, and extrapolate this data to be monthly instead
of yearly
We also extrapolate the data to 2019, 2020, and 2021 to match up with the Brazilian weather dataset.
We do it for each of the 4 columns: cumulative co2, co2, nitrous oxide, and methane.

Dec 31 of 1999 would be 11511.07836082474 <br>
Dec 31 of 2000 would be 11784.229247422683 <br>
Jan 31 of 2000 would be 11511.07836082474 + (dataFromCSV[1][0] - dataFromCSV[0][0]) / 12 <br>
Feb 31 of 2000 would be 11511.07836082474 + 2(dataFromCSV[1][0] - dataFromCSV[0][0]) / 12 <br>
Mar 31 of 2000 would be 11511.07836082474 + 3(dataFromCSV[1][0] - dataFromCSV[0][0]) / 12 <br>
.... <br>
.... <br>
Dec 31 of 2000 would be 11511.07836082474 + 12*(dataFromCSV[1][0] - dataFromCSV[0][0]) / 12 = 11784.229247422683 <br>

In [25]:
dataExtrapolatedJan2000ToDec2021 = []
year = 0
while year < 19:
    for month in range(12):
        dataExtrapolatedJan2000ToDec2021.append(dataFromCSV[year][0] + ((month+1) * (dataFromCSV[year+1][0] - dataFromCSV[year][0]) / 12))
    year += 1

In [254]:
# dataExtrapolatedJan2000ToDec2021 = []
# year = 0
# while year < 19:
#     for month in range(12):
#         dataExtrapolatedJan2000ToDec2021.append(dataFromCSV[year][1] + ((month+1) * (dataFromCSV[year+1][1] - dataFromCSV[year][1]) / 12))
#     year += 1

In [256]:
# dataExtrapolatedJan2000ToDec2021 = []
# year = 0
# while year < 19:
#     for month in range(12):
#         dataExtrapolatedJan2000ToDec2021.append(dataFromCSV[year][2] + ((month+1) * (dataFromCSV[year+1][2] - dataFromCSV[year][2]) / 12))
#     year += 1

In [258]:
# dataExtrapolatedJan2000ToDec2021 = []
# year = 0
# while year < 19:
#     for month in range(12):
#         dataExtrapolatedJan2000ToDec2021.append(dataFromCSV[year][3] + ((month+1) * (dataFromCSV[year+1][3] - dataFromCSV[year][3]) / 12))
#     year += 1

In [42]:
len(dataExtrapolatedJan2000ToDec2021) 

228

Now we currently have 19 years, from 2000-2018, and write the data to files to intermediate files to extrapolate for the years 2019, 2020, and 2021. We won't turn in these intermediate files unless requested.
We use a simple linear model (they are slightly nonlinear, so this is a source of error, but it was close enough that
a linear function was the best fit to extrapolate for our purposes).

In [43]:
with open('toExtrapolateCumulativeCO2v2.csv', 'w', newline='') as file:
    write = csv.writer(file)
    
    for row in dataExtrapolatedJan2000ToDec2021:
            write.writerow([row])

In [ ]:
# with open('toExtrapolateCO2v2.csv', 'w', newline='') as file:
#     write = csv.writer(file)
    
#     for row in dataExtrapolatedJan2000ToDec2021:
#             write.writerow([row])

In [ ]:
# with open('toExtrapolateNitrousOxidev2.csv', 'w', newline='') as file:
#     write = csv.writer(file)
    
#     for row in dataExtrapolatedJan2000ToDec2021:
#             write.writerow([row])

In [ ]:
# with open('toExtrapolateMethanev2.csv', 'w', newline='') as file:
#     write = csv.writer(file)
    
#     for row in dataExtrapolatedJan2000ToDec2021:
#             write.writerow([row])

Now the data is saved to csv files, and we have 22 years worth of data: from 2000 to 2021. We can combine each
of the 4 features into one big array to (later) combine with the weather dataset.

In [260]:
with open('toExtrapolateCumulativeCO2v2.csv', newline='') as readfile:
    reader = csv.reader(readfile)
    cumulativeCO2Extrapolation = tuple(reader)

In [261]:
with open('toExtrapolateCO2v2.csv', newline='') as readfile:
    reader = csv.reader(readfile)
    cO2Extrapolation = tuple(reader)

In [262]:
with open('toExtrapolateNitrousOxidev2.csv', newline='') as readfile:
    reader = csv.reader(readfile)
    nitrousOxideExtrapolation = tuple(reader)

In [263]:
with open('toExtrapolateMethanev2.csv', newline='') as readfile:
    reader = csv.reader(readfile)
    methaneExtrapolation = tuple(reader)

In [264]:
cumulativeCO2Extrapolation = [float(line[0]) for line in cumulativeCO2Extrapolation]
cO2Extrapolation = [float(line[0]) for line in cO2Extrapolation]
nitrousOxideExtrapolation = [float(line[0]) for line in nitrousOxideExtrapolation]
methaneExtrapolation = [float(line[0]) for line in methaneExtrapolation]

In [272]:
extrapolatedData = list(map(list, zip(cumulativeCO2Extrapolation, cO2Extrapolation, nitrousOxideExtrapolation, methaneExtrapolation)))
len(extrapolatedData)

264

In [273]:
extrapolatedData

[[11533.84093, 266.4969725, 27.66003866, 74.91455326],
 [11556.60351, 267.101878, 27.64806701, 74.89060137],
 [11579.36608, 267.7067835, 27.63609536, 74.86664948],
 [11602.12866, 268.311689, 27.62412371, 74.84269759],
 [11624.89123, 268.9165945, 27.61215206, 74.8187457],
 [11647.6538, 269.5215, 27.60018041, 74.79479381],
 [11670.41638, 270.1264055, 27.58820876, 74.77084192],
 [11693.17895, 270.731311, 27.57623711, 74.74689003],
 [11715.94153, 271.3362165, 27.56426546, 74.72293814],
 [11738.7041, 271.941122, 27.55229381, 74.69898625],
 [11761.46667, 272.5460275, 27.54032216, 74.67503436],
 [11784.22925, 273.150933, 27.52835052, 74.65108247],
 [11807.20466, 273.3637702, 27.54521048, 74.6382732],
 [11830.18008, 273.5766074, 27.56207045, 74.62546392],
 [11853.1555, 273.7894446, 27.57893041, 74.61265464],
 [11876.13091, 274.0022818, 27.59579038, 74.59984536],
 [11899.10633, 274.215119, 27.61265034, 74.58703608],
 [11922.08175, 274.4279562, 27.62951031, 74.5742268],
 [11945.05716, 274.640793

In [274]:
with open('allExtrapolatedGHGDataJan2000ToDec2021.csv', 'w', newline='') as file: 
    write = csv.writer(file)
    for monthYear in extrapolatedData:
        write.writerow(monthYear)

We save the 22 years x 12 months worth of data to a csv file to combine with the weather dataset in a separate Jupyter notebook.